src/Algebra/Graph.hs

```diff
@@ -1082,3 +1093,73 @@ sparsify graph = res
         m <- get
         put (m + 1)
         overlay <$> s `x` m <*> m `y` t
+
+{- Note [The rules of foldg]
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+The rules for foldg work like this.
+
+* Up to (but not including) phase 1, we use the "buildG/f" rule to
+  rewrite all saturated applications of f with its buildG/foldg
+  form, hoping for fusion to happen (through the "foldg/buildG" rule).
+
+  In phase 1 and 0, we switch off these rules, inline buildG, and
+  switch on the "graph/f" rule, which rewrites the "foldg/f"
+  thing back into plain map.
+
+  It's important that these two rules aren't both active at once
+  (along with build's unfolding) else we'd get an infinite loop
+  in the rules. Hence the activation control below.
+
+* FB functions are here to remember f after applying a "builG/f" rule.
+  These functions are higher-order functionsand therefore benefits from
+  inlining.
+
+* The "mapGFB" rule optimises compositions of mapG
+-}
+
+buildG :: forall a. F.Fold a -> Graph a
+buildG g = F.foldg Empty Vertex Overlay Connect g
+{-# INLINE [1] buildG #-}
+
+-- | FB functions
+mapGFB :: (b -> c) -> (a -> b) -> a -> c
+mapGFB = (.)
+{-# INLINE [0] mapGFB #-}
+
+induceFB :: b -> (a -> b) -> (a -> Bool) -> a -> b
```

Comment:
This function doesn't look like `induce` at all, so the name is a bit misleading.

Reply:
Perhaps, `matchFP` or something similar would be better.